### PR TITLE
Daniel

### PR DESCRIPTION
**Pull request by danielshulgin**

### CODE REVIEW

File: 2DSurviveGame/Assets/Scripts/Main/GameMeneger.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class GameMeneger : MonoBehaviour
```
Comment (arriven) [inline 2DSurviveGame/Assets/Scripts/Main/GameMeneger.cs]: Manager

File: 2DSurviveGame/Assets/Scripts/Player/Fire.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    Transform muzzleTransform;
+    [SerializeField]
+    InputSystem inputSystem;
+    void Start()
+    {
+        inputSystem.LeftButtonClick += Shot;
+    }
+    public void Shot()
+    {
+        playerMeneger.inventory.items.Find(item => item.itemType == ItemType.bullet).number -= 1;
```
Comment (arriven) [inline 2DSurviveGame/Assets/Scripts/Player/Fire.cs]: You need to check if you've found such object before accessing it.
also, I have a feeling that it would be more logical if the spawned prefab would be described by the inventory object rather than player itself. this would also split the responsibility better

File: 2DSurviveGame/Assets/Scripts/Player/Fire.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        GameObject bullet = SimplePool.Spawn(bulletPref, muzzleTransform.position, muzzleTransform.rotation);
+        BulletControl bulletControl = bullet.GetComponent<BulletControl>();
+        
+        bullet.GetComponent<Rigidbody2D>().AddForce(new Vector2(transform.up.x, transform.up.y) * bulletControl.speed);
```
Comment (arriven) [inline 2DSurviveGame/Assets/Scripts/Player/Fire.cs]: it would be better to get this from the input system rather than from transform. Less dependencies for the shooting process

File: 2DSurviveGame/Assets/Scripts/Player/FisicEngine.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    void Start()
+    {
+        inputSystem.LookDirection += Rotate;
+        inputSystem.Move += Move;
+    }
+
+    public void Move(Vector2 direction)
+    {
+        this.transform.position = this.transform.position + new Vector3(direction.x, direction.y, 0f) * speed;
+    }
```
Comment (arriven) [inline 2DSurviveGame/Assets/Scripts/Player/FisicEngine.cs]: the position difference should be affected by delta time

File: 2DSurviveGame/Assets/Scripts/Player/Fire.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+public class Fire : MonoBehaviour
+{
+    [SerializeField]
+    PlayerMeneger playerMeneger;
+    [SerializeField]
+    GameObject bulletPref;
+    [SerializeField]
+    Transform muzzleTransform;
+    [SerializeField]
+    InputSystem inputSystem;
```
Comment (arriven) [inline 2DSurviveGame/Assets/Scripts/Player/Fire.cs]: you should get input system from game object instead of inserting it as a field

File: 2DSurviveGame/Assets/Scripts/Main/InputSystem.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using System;
+
+public class InputSystem : MonoBehaviour
+{
+    public event Action<Vector2> Move;
+    public event Action<Vector2> LookDirection;
+    public event Action LeftButtonClick;
```
Comment (arriven) [inline 2DSurviveGame/Assets/Scripts/Main/InputSystem.cs]: I think it would be better to provide just a properties and use a polling system for input. It makes it easier to track what's happening with inputs

Comment (arriven) [reply 2DSurviveGame/Assets/Scripts/Main/InputSystem.cs]: also LeftButtonClick is too close to the implementation. Call it shoot or in case of polling IsShooting

**Review by arriven — CHANGES_REQUESTED** [review]

File: 2DSurviveGame/Assets/Scripts/Player/Fire.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    IEnumerator Shoot()
+    {
+        while (true)
+        {
+            //TODO
+            if (!OneShoot())
+            {
+                break;
+            }
+            yield return new WaitForSeconds(.5f);
```
Comment (arriven) [inline 2DSurviveGame/Assets/Scripts/Player/Fire.cs:82]: avoid hardcoded values

File: 2DSurviveGame/Assets/Scripts/Main/FlowController.cs [hunk 8]
```diff
@@ -0,0 +1,8 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class GameMeneger : MonoBehaviour
```
Comment (arriven) [inline 2DSurviveGame/Assets/Scripts/Main/FlowController.cs]: still meneger -> manager

File: 2DSurviveGame/Assets/Scripts/Entity/Inventory.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        return number - n > 0;
+    }
+}
+
+//TODO
+//public class Bullet : Item
+//{
+//    public Bullet(ItemType itemType, int number = 1) : base(itemType, number)
+//    {
+//        this.itemType = itemType;
```
Comment (arriven) [inline 2DSurviveGame/Assets/Scripts/Entity/Inventory.cs]: even if this is not in the build yet, you don't need to initialize this values in derived class. they are already initiated by the base class constructor

File: 2DSurviveGame/Assets/Scripts/Player/Fire.cs [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        input = new InputManager();
+    }
+
+    private void OnEnable()
+    {
+        input.Player.StartShoot.performed += context => StartShooting();
+        input.Player.EndShoot.performed += context => StopShooting();
+        input.Player.StartShoot.Enable();
+        input.Player.EndShoot.Enable();
```
Comment (arriven) [inline 2DSurviveGame/Assets/Scripts/Player/Fire.cs]: is it possible to encapsulate these 2 calls per callback into one?

File: 2DSurviveGame/Assets/Scripts/Player/Fire.cs [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        input.Player.EndShoot.Disable();
+    }
+
+    public bool OneShoot()
+    {
+        Item item = playerManager.inventory.items.Find(_item => _item.itemType == ItemType.Bullet);
+        if (item != null && item.Use())
+        {
+            GameObject bullet = Instantiate(bulletPref, muzzleTransform.position, muzzleTransform.rotation);
+            BulletControl bulletControl = bullet.GetComponent<BulletControl>();
```
Comment (arriven) [inline 2DSurviveGame/Assets/Scripts/Player/Fire.cs]: why is this line needed?

File: 2DSurviveGame/Assets/Scripts/Player/FisicEngine.cs [hunk 12]
```diff
@@ -0,0 +1,47 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Experimental.Input;
+
+public class FisicEngine : MonoBehaviour
```
Comment (arriven) [inline 2DSurviveGame/Assets/Scripts/Player/FisicEngine.cs]: fisic -> physics

**Review by arriven — CHANGES_REQUESTED** [review]

File: 2DSurviveGame/Assets/Scripts/Player/Fire.cs (hunk 7 above)
Comment (danielshulgin) [reply 2DSurviveGame/Assets/Scripts/Player/Fire.cs:82]: I wanted to introduce more complex functionality in the future to get the frequency of shots from weapon parameters.
